### PR TITLE
doc: document process for access to jenkins jobs

### PR DESCRIPTION
**Pull request by mhdawson**
Document process for allowing members outside
of the build working group the access required to be able
to edit or run jobs associated with their working group.

### CODE REVIEW

**Comment by mhdawson** [comment]
@nodejs/build please take a look.
Supports: https://github.com/nodejs/build/issues/526

File: doc/process/jenkins_job_configuation_access.md [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+In the first case we want to be able to provide ongoing access to
+an individual to edit or run a job associated with a working group
+when this supports the efforts of a working group/team. For
+example, some members of the benchmarking team will be regularly
+modifying benchmarking jobs and therefore the ability to configure
+benchmarking jobs needs to be granted in an ongoing manner.
+
+In these cases the following will be used to decide if ongoing access
+can be provided:
```
Comment (gibfahn) [inline doc/process/jenkins_job_configuation_access.md]: It seems odd to have to go through this process for each member of a Working Group. It also seems odd for the Build team to be approving WG members' access to a WG's jobs. Wouldn't it be easier to say that (for example) the Benchmarking team own the benchmarking jobs, i.e. being a member of the WG gives access to that WG's jobs?

I think this _**may**_ be what @jbergstroem was saying in https://github.com/nodejs/build/issues/526#issuecomment-261390899

Comment (jbergstroem) [inline doc/process/jenkins_job_configuation_access.md]: @gibfahn yeah; that's pretty much my thoughts as well. The possible distinction would be that each team can have an 'admin' group if the team is big so not all have free reign.

Comment (gibfahn) [inline doc/process/jenkins_job_configuation_access.md]: @jbergstroem Yeah sure, that'd make sense, but presumably that'd be something the WG would manage themselves?

Comment (jbergstroem) [inline doc/process/jenkins_job_configuation_access.md]: @gibfahn yeah

Comment (jbergstroem) [inline doc/process/jenkins_job_configuation_access.md]: It's at least my preferred way; we should really delegate job control to people that works on it. If the group is big enough (or a trust issue) to separate members (execute and admins (edit), that should be preferred.

Comment (gibfahn) [inline doc/process/jenkins_job_configuation_access.md]: So maybe the easiest way would be to have a process (raise an issue in this repo) for giving a WG access to a group of jobs, and then have the WG agree on an initial list of members to have access, which could be everyone or a subgroup. If it is a subgroup there should probably be a separate @nodejs team called `jenkins-[wg-name]`.

Once there is an initial team, then the WG can manage ongoing maintenance (adding/removing people) themselves.

After that I think the only thing that would need Build team help would be creating new jobs, which would probably be just asking someone in @nodejs/jenkins-admins to clone one of the existing jobs (which should maintain permissions).

File: doc/process/jenkins_job_configuation_access.md [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+* individual
+* collaborator status
+* information supporting request as per the items listed above
+* run/cancel requested: yes/no
+* edit requested: yes/no
+* Period requested for:  Indefinite or a specific from -> to
+
+The request should start with 2 lgtms from working group members and
+then the the build WG will review and vote on the request.
```
Comment (joaocgreis) [inline doc/process/jenkins_job_configuation_access.md]: About the build WG approval, I don't think we need a formal vote if there are no objections. I'd be ok with having the issue open for the standard 48/72 and if there are no objections then move forward. This could be longer (1 week?) if anyone would be more comfortable.

To create new jobs, if the person who is asking for a clone has permission to edit the original job, requests should be immediately granted. So, both IRC or issue are ok. Same for new blank jobs, since we can just delete everything from a clone of the same type.

**Comment by mhdawson** [comment]
@jbergstroem, @gibfahn @joaocgreis updated based on comments so far.  Let me know what you think.

File: doc/process/jenkins_job_configuation_access.md [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+This document outlines the processes by which we will allow
+people outside the build working group to edit jobs as well
+as allowing people who are not collaborators run specific jobs.
+
+Unfortunately jenkins does not allow us to easily delegate the
+creation of new jobs in an appropriate manner.
+As such, the creation of a new job will
+need to be requested through an issue in the build repo.  Any
+build member can then create a blank job in the group for the
```
Comment (gibfahn) [inline doc/process/jenkins_job_configuation_access.md]: Maybe `create a blank job` -> `clone one of that workgroup's existing jobs`, as the cloning will then automatically pick up the right privileges.

File: doc/process/jenkins_job_configuation_access.md [hunk 4]
```diff
@@ -0,0 +1,72 @@
+# Introduction
+
+There are a number of cases were we would like working group members
```
Comment (gibfahn) [inline doc/process/jenkins_job_configuation_access.md]: s/were/where

File: doc/process/jenkins_job_configuation_access.md [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+  in some other way ?
+* Are there collaborators who work with the individual and can vouch
+  for them.
+
+It is suggested that this consideration be documented in an
+issue on the workgroups repo.
+
+Once approved by the working group and existing workgroup admins, one
+of the existing workgroup admins can then add the new individual to
+the github admin team for the workgroup (ex benchmarking-admins).
```
Comment (gibfahn) [inline doc/process/jenkins_job_configuation_access.md]: Maybe `e.g.` rather than `ex`?

File: doc/process/jenkins_job_configuation_access.md [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+A subset of the members of a workgroup will be able to modify
+and delete jobs. This will be accomplished by enabling
+project-based security for the jobs and adding the group
+corresponding to the github team for the working group admins
+where there Build/Cancel/Discover/Read/Delete/Update
+rights have been set.
+
+Since the ability to modify jobs opens up new attack vectors,
+we ask that the workgroups limit this access to those
+individuals that require it, and when appropriate the length,
```
Comment (gibfahn) [inline doc/process/jenkins_job_configuation_access.md]: Comma after length is unneeded.

File: doc/process/jenkins_job_configuation_access.md [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+should be considered:
+
+* Does the scope and size of the need justify providing access.
+* Is the invidual a Node.js collaborator? If so then access should
+  be allowed provided the first point is satisfied.
+* Length and consistency of involvement with Node.js working groups
+  and/or community.
+* Consequences to the invidudal in case of mis-behaviour. For example,
+  would they potentially lose their job if they were reported as
+  mis-behaving to their employer ? Would being banned from involvement
```
Comment (gibfahn) [inline doc/process/jenkins_job_configuation_access.md]: Space before `?`

File: doc/process/jenkins_job_configuation_access.md [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+  in the Node.js community negatively affect them personally
+  in some other way ?
```
Comment (gibfahn) [inline doc/process/jenkins_job_configuation_access.md]: Space before `?` again

**Comment by mhdawson** [comment]
@gibfahn @jbergstroem  second set of comments addressed.

File: doc/process/jenkins_job_configuation_access.md [hunk 9]
```diff
@@ -0,0 +1,74 @@
+# Introduction
+
+There are a number of cases where we would like working group members
+to be able to run and or edit jobs within jenkins. Examples include
+the CITGM jobs and the Benchmarking jobs.
+
+By default jobs can only be edited by members of the
+'nodejs/jenkins-admins' group which contains a subset of the build
```
Comment (gibfahn) [inline doc/process/jenkins_job_configuation_access.md]: `[@nodejs/jenkins-admins](https://github.com/orgs/nodejs/teams/jenkins-admins)`

File: doc/process/jenkins_job_configuation_access.md [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+workgroup members. We do not give access to
+everbody to run/edit as it would be a potential security issue,
```
Comment (gibfahn) [inline doc/process/jenkins_job_configuation_access.md]: s/everbody/everybody

File: doc/process/jenkins_job_configuation_access.md [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+both through the potential disclosure of secrets used as
+part of the builds, as well as a potential denial
+of service if jobs are launched maliciously.
+
+This document outlines the processes by which we will allow
+people outside the build working group to edit jobs as well
+as allowing people who are not collaborators run specific jobs.
```
Comment (gibfahn) [inline doc/process/jenkins_job_configuation_access.md]: run -> to run

File: doc/process/jenkins_job_configuation_access.md [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+In the case of granting access to edit jobs the following
+should be considered:
+
+* Does the scope and size of the need justify providing access.
+* Is the invidual a Node.js collaborator? If so then access should
+  be allowed provided the first point is satisfied.
+* Length and consistency of involvement with Node.js working groups
+  and/or community.
+* Consequences to the invidudal in case of mis-behaviour. For example,
+  would they potentially lose their job if they where reported as
```
Comment (gibfahn) [inline doc/process/jenkins_job_configuation_access.md]: s/where/were

File: doc/process/jenkins_job_configuation_access.md [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+enabled.
+
+## Ability to modify jobs
+
+A subset of the members of a workgroup will be able to modify
+and delete jobs. This will be accomplished by enabling
+project-based security for the jobs and adding the group
+corresponding to the github team for the working group admins
+where there Build/Cancel/Discover/Read/Delete/Update
+rights have been set.
```
Comment (gibfahn) [inline doc/process/jenkins_job_configuation_access.md]: >This will be accomplished by enabling project-based security for the jobs and adding the group corresponding to the github team for the working group admins where there Build/Cancel/Discover/Read/Delete/Update rights have been set.

A bit hard to parse, maybe:

This will be accomplished by enabling project-based security for the jobs, and then giving the jenkins group corresponding to the github team for the workgroup admins Build/Cancel/Discover/Read/Delete/Update permissions.

**Comment by mhdawson** [comment]
Addressed next set comments.

File: doc/process/jenkins_job_configuation_access.md [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+Since the ability to modify jobs opens up new attack vectors,
+we ask that the workgroups limit this access to those
+individuals that require it, and when appropriate, the length
+of time access is granted.
+
+In the case of granting access to edit jobs the following
+should be considered:
+
+* Does the scope and size of the need justify providing access.
+* Is the invidual a Node.js collaborator? If so then access should
```
Comment (gibfahn) [inline doc/process/jenkins_job_configuation_access.md]: s/invidual/individual

File: doc/process/jenkins_job_configuation_access.md [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+  be allowed provided the first point is satisfied.
+* Length and consistency of involvement with Node.js working groups
+  and/or community.
+* Consequences to the invidudal in case of mis-behaviour. For example,
```
Comment (gibfahn) [inline doc/process/jenkins_job_configuation_access.md]: s/invidudal/individual

File: doc/process/jenkins_job_configuation_access.md [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+needed, a build member can clone one of the workgroup's 
+existing jobs, which will preserve the right permissions.
+
+## Ability to run/cancel jobs
+
+All members of a workgroup will be able to run and cancel
+the jobs tied to their workgroup. This will be accomplished
+by enabling project-based security for the jobs, and then
+giving the jenkins group corresponding to the github
+team for the workgroup admins
```
Comment (gibfahn) [inline doc/process/jenkins_job_configuation_access.md]: I think you modified the wrong paragraph here (see https://github.com/nodejs/build/pull/554#discussion_r91430337).

File: doc/process/jenkins_job_configuation_access.md (hunk 15 above)
Comment (mhdawson) [inline doc/process/jenkins_job_configuation_access.md]: That word must have been hard for me fixed.

**Comment by mhdawson** [comment]
Addressed new comments.

**Comment by mhdawson** [comment]
@jbergstroem can you review now that @gibfahn is happy with it.

File: doc/process/jenkins_job_configuation_access.md [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+and delete jobs.  This will be accomplished
+by enabling project-based security for the jobs, and then
+giving the jenkins group, corresponding to the github
+team for the workgroup admins,
+the Build/Cancel/Discover/Read/Delete/Update permissions.
+
+Since the ability to modify jobs opens up new attack vectors,
+we ask that the workgroups limit this access to those
+individuals that require it, and when appropriate, the length
+of time access is granted.
```
Comment (jbergstroem) [inline doc/process/jenkins_job_configuation_access.md]: We should also mention the fact that our resources aren't unlimited. Spawning long-running jobs on arm (or even windows) can easily pile up.

**Comment by gibfahn** [comment]
@mhdawson After https://github.com/nodejs/security-wg/pull/9#discussion_r92052847, maybe we should `s/workgroup/working group` everywhere...

**Comment by mhdawson** [comment]
Ok address most of the remaining comments.  @jbergstroem in respect to formatting to W80 is there an easy way to do that ?  I've tried to stick under the 72 chars but evening out by hand would be a bit of work and I don't think it matters since when its viewed in markdown, markdown will handle making it look good.

**Comment by jbergstroem** [comment]
@mhdawson: 72 or 80 doesn't really matter. For me, the point of markdown is that its very readable rendered (github) and in your editor (vim, et al). A consistent line length is just generally easier to consume. Anyway, since we're down to bikeshedding I'm `+1` regardless.

**Comment by gibfahn** [comment]
@nodejs/build would be really good to get this landed so people in teams can get access to jobs.

@mhdawson With two reviews I think if there are no objections this should be good to land, people can always PR improvements.

File: doc/process/jenkins_job_configuation_access.md [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+people outside the build working group to edit jobs as well
+as allowing people who are not collaborators to run specific jobs.
+
+Unfortunately jenkins does not allow us to easily delegate the
+creation of new jobs in an appropriate manner.
+As such, the creation of a new job will
+need to be requested through an issue in the build repo. Any
+build member can then create a blank job in the group for the
+working group which can then be edited by those with access
+to the jobs for that working group. As and when more jobs are
```
Comment (phillipj) [inline doc/process/jenkins_job_configuation_access.md]: > As and when more jobs are ..

Something wrong with that sentence?

**Comment by phillipj** [comment]
Other than a microscopic nit, this LGTM

File: doc/process/jenkins_job_configuation_access.md (hunk 18 above)
Comment (gibfahn) [inline doc/process/jenkins_job_configuation_access.md]: Looks okay to me, `As and when` is a (weird) English phrase

Comment (phillipj) [inline doc/process/jenkins_job_configuation_access.md]: Hm alright, ignore my comment then

**Comment by mhdawson** [comment]
Since there are 2 LGTMs now I'll plan to land tomorrow.

**Comment by gibfahn** [comment]
Landed in https://github.com/nodejs/build/commit/1fcbba2ae6242588a1f7cb779a859d8734213097
